File: Task/Pfaffian-and-hfaffian/Python/pfaffian-and-hfaffian.py
from itertools import permutations
from operator import mul
from math import fsum, factorial
from functools import reduce
from operator import itemgetter

DEBUG = False  # like the built-in __debug__


def spermutations(n):
    """permutations by swapping. Yields: perm, sign"""
    sign = 1
    p = [[i, 0 if i == 0 else -1]  # [num, direction]
         for i in range(n)]

    if DEBUG: print(' #', p)
    yield tuple(pp[0] for pp in p), sign

    while any(pp[1] for pp in p):  # moving
        i1, (n1, d1) = max(((i, pp) for i, pp in enumerate(p) if pp[1]),
                           key=itemgetter(1))
        sign *= -1
        if d1 == -1:
            # Swap down
            i2 = i1 - 1
            p[i1], p[i2] = p[i2], p[i1]
            # If this causes the chosen element to reach the First or last
            # position within the permutation, or if the next element in the
            # same direction is larger than the chosen element:
            if i2 == 0 or p[i2 - 1][0] > n1:
                # The direction of the chosen element is set to zero
                p[i2][1] = 0
        elif d1 == 1:
            # Swap up
            i2 = i1 + 1
            p[i1], p[i2] = p[i2], p[i1]
            # If this causes the chosen element to reach the first or Last
            # position within the permutation, or if the next element in the
            # same direction is larger than the chosen element:
            if i2 == n - 1 or p[i2 + 1][0] > n1:
                # The direction of the chosen element is set to zero
                p[i2][1] = 0
        if DEBUG: print(' #', p)
        yield tuple(pp[0] for pp in p), sign

        for i3, pp in enumerate(p):
            n3, d3 = pp
            if n3 > n1:
                pp[1] = 1 if i3 < i2 else -1
                if DEBUG: print(' # Set Moving')


def prod(lst):
    return reduce(mul, lst, 1)


def is_antisymmetric(a):
    """Check if a matrix is antisymmetric"""
    n = len(a)
    for i in range(n):
        if a[i][i] != 0:  # Diagonal elements must be zero
            return False
        for j in range(i + 1, n):
            if a[i][j] != -a[j][i]:
                return False
    return True


def is_symmetric(a):
    """Check if a matrix is symmetric"""
    n = len(a)
    for i in range(n):
        for j in range(i + 1, n):
            if a[i][j] != a[j][i]:
                return False
    return True


def pfaffian(a):
    """Compute the Pfaffian of a 2n x 2n antisymmetric matrix."""
    n = len(a) // 2

    if len(a) % 2 != 0:
        raise ValueError("Matrix must be 2n x 2n for Pfaffian.")

    if not is_antisymmetric(a):
        raise ValueError("The Pfaffian does not support Non-antisymmetric matrices yet.")

    # Calculate normalization factor
    normalization = 1.0 / (2 ** n * factorial(n))

    indices = range(2 * n)
    result = fsum(
        sign * prod(a[sigma[2 * i]][sigma[2 * i + 1]] for i in range(n))
        for sigma, sign in spermutations(2 * n)
    )

    # Apply the normalization factor
    return normalization * result


def hfaffian(a):
    """Compute the hfaffian of a 2n x 2n symmetric matrix."""
    n = len(a) // 2

    if len(a) % 2 != 0:
        raise ValueError("Matrix must be 2n x 2n for hfaffian.")

    if not is_antisymmetric(a):
        raise ValueError("The Hfaffian does not support Non-antisymmetric matrices yet.")

    # Calculate normalization factor
    normalization = 1.0 / (2 ** n * factorial(n))

    indices = range(2 * n)
    result = fsum(
        prod(a[sigma[2 * i]][sigma[2 * i + 1]] for i in range(n))
        for sigma in permutations(indices)
    )

    # Apply the normalization factor
    return normalization * result


if __name__ == '__main__':
    from pprint import pprint as pp

    for a in (
            [
                [0, 1, -1, 2],
                [-1, 0, 3, -4],
                [1, -3, 0, 5],
                [-2, 4, -5, 0]],  # Example of an antisymmetric matrix for Pfaffian

            [
                [1, 2, 3, 4, 5, 6],
                [2, 7, 8, 9, 10, 11],
                [3, 8, 12, 13, 14, 15],
                [4, 9, 13, 16, 17, 18],
                [5, 10, 14, 17, 19, 20],
                [6, 11, 15, 18, 20, 21]],  # Example of a symmetric matrix for hfaffian
    ):
        print('')
        pp(a)
        if len(a) % 2 == 0:
            try:
                print('Pfaffian: %s' % pfaffian(a))
            except ValueError as e:
                print('Pfaffian Error:', e)
            try:
                print('Hfaffian: %s' % hfaffian(a))
            except ValueError as e:
                print('Hfaffian Error:', e)
        else:
            print('Matrix size must be even for Pfaffian and hfaffian.')

File: Task/Pfaffian-and-hfaffian/C-sharp/pfaffian-and-hfaffian.cs
using System;
using System.Collections.Generic;

public class PfaffianAndHfaffian
{
    public static void Main(string[] args)
    {
        List<int[,]> matrices = new List<int[,]>
        {
            new int[,] { {  0,  1 }, { -1,  0 } }, // Tiny 2 x 2 matrix

            new int[,] { {  0,  1, -1,  2 }, // Small 4 x 4 matrix
                          { -1,  0,  3, -4 },
                          {  1, -3,  0,  5 },
                          { -2,  4, -5,  0 } },

            new int[,] { { 1,  2,  3,  4,  5,  6 }, // Symmetric 6 x 6 matrix
                          { 2,  7,  8,  9, 10, 11 },
                          { 3,  8, 12, 13, 14, 15 },
                          { 4,  9, 13, 16, 17, 18 },
                          { 5, 10, 14, 17, 19, 20 },
                          { 6, 11, 15, 18, 20, 21 } },

            new int[,] { {  0,  1,  2,  3,  4,  5,  6,  7,  8,  9 }, // Larger 10 x 10 matrix
                          { -1,  0,  8,  7,  6,  5,  4,  3,  2,  1 },
                          { -2, -8,  0,  1,  2,  3,  4,  5,  6,  7 },
                          { -3, -7, -1,  0,  6,  5,  4,  3,  2,  1 },
                          { -4, -6, -2, -6,  0,  1,  2,  3,  4,  5 },
                          { -5, -5, -3, -5, -1,  0,  4,  3,  2,  1 },
                          { -6, -4, -4, -4, -2, -4,  0,  1,  2,  3 },
                          { -7, -3, -5, -3, -3, -3, -1,  0,  2,  1 },
                          { -8, -2, -6, -2, -4, -2, -2, -2,  0,  1 },
                          { -9, -1, -7, -1, -5, -1, -3, -1, -1,  0 } }
        };

        foreach (var matrix in matrices)
        {
            PrintMatrix(matrix);
            foreach (Faffian faffian in Enum.GetValues(typeof(Faffian)))
            {
                var result = ComputeFaffian(matrix, faffian);
                if (result.HasValue)
                {
                    Console.WriteLine($"{faffian}: {result.Value}");
                }
            }
            Console.WriteLine();
        }
    }

    private static long? ComputeFaffian(int[,] matrix, Faffian faffian)
    {
        int size = matrix.GetLength(0);
        if (size % 2 != 0)
        {
            Console.WriteLine($"Matrix size must be even for {faffian}");
            return null;
        }

        if (!IsAntisymmetric(matrix))
        {
            Console.WriteLine($"The {faffian} does not support non-antisymmetric matrices");
            return null;
        }

        int n = size / 2;
        int sum = 0;
        List<SignedPerm> signedPerms = SignedPermutations(2 * n - 1);
        foreach (var signedPerm in signedPerms)
        {
            List<int> sigma = signedPerm.Permutation;
            int sign = (faffian == Faffian.Pfaffian) ? signedPerm.Sign : 1;
            int product = 1;
            for (int i = 0; i < n; i++)
            {
                product *= matrix[sigma[2 * i], sigma[2 * i + 1]];
            }
            sum += sign * product;
        }

        double normalization = 1.0 / Factorial(n) / Math.Pow(2, n);
        return (long)Math.Round(sum * normalization);
    }

    private static List<SignedPerm> SignedPermutations(int n)
    {
        List<int> perms = new List<int>();
        for (int i = 0; i <= n; i++)
        {
            perms.Add(i);
        }
        List<SignedPerm> signedPerms = new List<SignedPerm> { new SignedPerm(new List<int>(perms), 1) };
        int sign = 1;
        for (int k = 1; k < Factorial(n + 1); k++)
        {
            int i = n - 1;
            int j = n;
            while (perms[i] > perms[i + 1])
            {
                i--;
            }
            while (perms[j] < perms[i])
            {
                j--;
            }
            Swap(perms, i, j);
            sign = -sign;
            i++;
            j = n;
            while (i < j)
            {
                Swap(perms, i, j);
                sign = -sign;
                i++;
                j--;
            }
            signedPerms.Add(new SignedPerm(new List<int>(perms), sign));
        }
        return signedPerms;
    }

    private static bool IsAntisymmetric(int[,] matrix)
    {
        int size = matrix.GetLength(0);
        for (int i = 0; i < size; i++)
        {
            if (matrix[i, i] != 0)
            {
                return false;
            }
            for (int j = i + 1; j < size; j++)
            {
                if (matrix[i, j] != -matrix[j, i])
                {
                    return false;
                }
            }
        }
        return true;
    }

    private static int Factorial(int n)
    {
        int factorial = 1;
        for (int i = 2; i <= n; i++)
        {
            factorial *= i;
        }
        return factorial;
    }

    private static void Swap(List<int> list, int i, int j)
    {
        int temp = list[i];
        list[i] = list[j];
        list[j] = temp;
    }

    private static void PrintMatrix(int[,] matrix)
    {
        for (int i = 0; i < matrix.GetLength(0); i++)
        {
            Console.Write("|");
            for (int j = 0; j < matrix.GetLength(1); j++)
            {
                Console.Write($"{matrix[i, j],2}, ");
            }
            Console.WriteLine();
        }
    }

    private class SignedPerm
    {
        public List<int> Permutation { get; }
        public int Sign { get; }

        public SignedPerm(List<int> permutation, int sign)
        {
            Permutation = permutation;
            Sign = sign;
        }
    }

    private enum Faffian { Pfaffian, Hfaffian }
}

